qmt/geometry/geo_2d_data.py

```diff
@@ -132,10 +134,63 @@ def coord_list(self, part_name: str) -> List:
         elif isinstance(part, LineString):
             return list(np.array(part.coords.xy).T)[:]
 
+    def crop(
+        self,
+        x_min: float = None,
+        x_max: float = None,
+        y_min: float = None,
+        y_max: float = None,
+    ):
+        if x_min is not None and x_max is not None:
+            assert x_min < x_max, "x_max must be greater than x_min"
+        if y_min is not None and y_max is not None:
+            assert y_min < y_max, "y_max must be greater than y_min"
+        cropped_geo = Geo2DData(self.lunit)
+        x_min_old, x_max_old, y_min_old, y_max_old = self.compute_bb()
+        x_min = x_min_old if x_min is None else x_min
+        x_max = x_max_old if x_max is None else x_max
+        y_min = y_min_old if y_min is None else y_min
+        y_max = y_max_old if y_max is None else y_max
+        crop_poly = Polygon(
+            [[x_min, y_min], [x_min, y_max], [x_max, y_max], [x_max, y_min]]
+        )
+
+        ind_count = defaultdict(int)
+
+        for name, part in self.parts.items():
+            ind_name = name.split(":")[0] if ":" in name else name
+            cropped_part = crop_poly.intersection(part)
+            if cropped_part.is_empty:
+                continue
+            if isinstance(cropped_part, Polygon) or isinstance(
+                cropped_part, LineString
+            ):
+                cropped_geo.add_part(name, cropped_part)
+                ind_count[ind_name] += 1
+            elif isinstance(cropped_part, GeometryCollection):
```

Comment:
At initialization, Geo2DData is intended to made from a Polygon or LineString, do we want to introduce other geometry types?

Reply:
This is necessary because the intersection of 2 Polygon/LineString objects may not return a Polygon/LineString . Think about this:
![untitled](https://user-images.githubusercontent.com/4095626/59541968-94051b00-8eb8-11e9-85f5-6d588a805aef.png)
the intersection of crop and area1 will be two polygons, and it's returned as a GeometryCollection. In this case we iterate through it and add each polygon (as area1:0 and area1:1)